app/src/main/kotlin/de/rochefort/childmonitor/MonitorService.kt

```diff
@@ -0,0 +1,270 @@
+                val ma = monitorActivity
+                ma?.runOnUiThread {
+                    val statusText = ma.findViewById<TextView>(R.id.textStatus)
+                    statusText.setText(R.string.waitingForParent)
+                    val serviceText = ma.findViewById<TextView>(R.id.textService)
+                    serviceText.text = serviceName
+                    val portText = ma.findViewById<TextView>(R.id.port)
+                    portText.text = Integer.toString(port)
+                }
+            }
+
+            override fun onRegistrationFailed(serviceInfo: NsdServiceInfo, errorCode: Int) {
+                // Registration failed!  Put debugging code here to determine why.
+                Log.e(TAG, "Registration failed: $errorCode")
+            }
+
+            override fun onServiceUnregistered(arg0: NsdServiceInfo) {
+                // Service has been unregistered.  This only happens when you call
+                // NsdManager.unregisterService() and pass in this listener.
+                Log.i(TAG, "Unregistering service")
+            }
+
+            override fun onUnregistrationFailed(serviceInfo: NsdServiceInfo, errorCode: Int) {
+                // Unregistration failed.  Put debugging code here to determine why.
+                Log.e(TAG, "Unregistration failed: $errorCode")
+            }
+        }
+        nsdManager.registerService(
+                serviceInfo, NsdManager.PROTOCOL_DNS_SD, registrationListener)
+    }
+
+    private fun unregisterService() {
+        this.registrationListener?.let {
+            this.registrationListener = null
+            Log.i(TAG, "Unregistering monitoring service")
+            this.nsdManager.unregisterService(it)
+        }
+    }
+
+    private fun createNotificationChannel() {
+        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O) {
+            val serviceChannel = NotificationChannel(
+                    CHANNEL_ID,
+                    "Foreground Service Channel",
+                    NotificationManager.IMPORTANCE_DEFAULT
+            )
+            this.notificationManager.createNotificationChannel(serviceChannel)
+        }
+    }
+
+    private fun buildNotification(): Notification {
+        val text: CharSequence = "Child Device"
+        // Set the info for the views that show in the notification panel.
+        val b = NotificationCompat.Builder(this, CHANNEL_ID)
+        b.setSmallIcon(R.drawable.listening_notification) // the status icon
+                .setOngoing(true)
+                .setTicker(text) // the status text
+                .setContentTitle(text) // the label of the entry
+        return b.build()
+    }
+
+    override fun onDestroy() {
+        this.monitorThread?.let {
+            this.monitorThread = null
+            it.interrupt()
+        }
+        unregisterService()
+        this.connectionToken = null
+        this.currentSocket?.let {
+            this.currentSocket = null
```

Comment:
also here: move this line out of the lambda. it is always okay to set this to null.

Reply:
Done.